Add TablePropertiesCollector support

pcholakov opened the pull request:
Stack:

- #8
- __->__ #9 ([relevant commits only](https://github.com/restatedev/rust-rocksdb/pull/9/files/116663b9243431f8b5b70db794baa30657d5153e..4c4e1114541fd4bcf40f30cf2b482bb94d93d9be))

Depends on: https://github.com/restatedev/rocksdb/pull/8

---

Additional testing:

```
pavel@ubuntu:/Users/pavel/restate/rust-rocksdb$ cargo valgrind nextest run --no-capture event_listener table_properties
   Compiling rust-librocksdb-sys v0.36.0+10.1.3 (/Users/pavel/restate/rust-rocksdb/librocksdb-sys)
   Compiling rust-rocksdb v0.40.0-restate.1 (/Users/pavel/restate/rust-rocksdb)
    Finished `test` profile [unoptimized + debuginfo] target(s) in 1m 09s
info: using target runner `/home/pavel/.cargo/bin/cargo-valgrind` defined by environment variable `CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_RUNNER`(test), rust_rocksdb(test), test_write_────────────
 Nextest run ID 68c1ec12-c06b-422c-86b9-f18d8c07dc67 with nextest profile: default
    Starting 2 tests across 25 binaries (195 tests skipped)
       START             rust-rocksdb::test_event_listener test_event_listener

running 1 test
test test_event_listener ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.69s

        PASS [   3.348s] rust-rocksdb::test_event_listener test_event_listener
       START             rust-rocksdb::test_table_properties test_table_properties_collector

running 1 test
test test_table_properties_collector ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.70s

        PASS [   3.207s] rust-rocksdb::test_table_properties test_table_properties_collector
────────────
     Summary [   6.556s] 2 tests run: 2 passed, 195 skipped
```

pcholakov commented on src/table_properties.rs:
```diff
@@ -0,0 +1,315 @@
+        entry_type: EntryType,
+        seq: u64,
+        file_size: u64,
+    ) -> bool;
```
Something like `Result<(), Err>` might be more appropriate here. I'll see what feels natural as I build the LSN tracking.

pcholakov replied:
Having looked at how RocksDB uses the returned `Status` from the native factory interface, I don't think this would add any value.

tillrohrmann replied:
How does RocksDB use the returned `Status`?

pcholakov replied:
C++ callsite: https://github.com/facebook/rocksdb/blob/ab8bd8455a055c26885c5f16734c816d379dc669/table/meta_blocks.cc#L200-L207

Related thread from C FFI PR: https://github.com/restatedev/rocksdb/pull/8/commits/805888c4babad622aee85dceb7518ffaeca4b010#r2066640064

AhmedSoliman commented on src/table_properties.rs:
```diff
@@ -0,0 +1,300 @@
+            let value = if value_ptr.is_null() {
+                None
+            } else {
+                let value_string = CStr::from_ptr(value_ptr).to_str().unwrap().to_owned();
```
Are you sure this is a utf-8 string?

pcholakov replied:
Absolutely not! I've updated the property collector interface to be `HashMap<CString, CString>` to make this clearer from both reading and writing these.

pcholakov committed:
Add TablePropertiesCollector support

This change the ability to configure custom TablePropertiesCollector
factories, and to access user-collected properties via the FlushJobInfo
event.

pcholakov committed:
Switch all user property interfaces to use CString

pcholakov commented on src/event_listener.rs:
```diff
@@ -114,6 +117,12 @@ pub struct FlushJobInfo {
+
+    // holds a pointer to data borrowed from RocksDB for the duration of the event handler callback;
+    // we are responsible for freeing the wrapper struct
+    table_properties: NonNull<ffi::rocksdb_table_properties_t>,
```
I chose to model it like this to avoid wholesale copying of the user properties map - and there is quite a lot else going on inside the native `TableProperties` struct though mostly primitive properties. Having the C wrapper handy allows us to access just the things we need. Unless we make a deep copy of the property dictionary upfront, we'll always need two derefs to get to a specific property value by key, either via `FlushJobInfo->TableProperties->UserCollectedProperties` or (as implemented) `rocksdb_table_properties_t->TableProperties->UserCollectedProperties`. In an earlier iteration, I was holding on to a `UserProperties` pointer here, but it seemed like `TableProperties` is the right level as it also contains some primitives that might be useful to expose in the future.

https://github.com/restatedev/rocksdb/blob/ca02f21519a0e006cef2b9c3bfa8e955ead90941/include/rocksdb/table_properties.h#L212

pcholakov commented on src/table_properties.rs:
```diff
@@ -0,0 +1,244 @@
+
+/// Extension trait for [`Options`] to register table properties collectors
+pub trait TablePropertiesExt {
+    fn add_table_properties_collector_factory<F>(&mut self, factory_fn: F)
```
A potential simplification to the API is to replace the trait with an owned `CString` name and `FnMut(TablePropertiesCollectorContext) -> TablePropertiesCollector`. I thought it wasn't that much of an improvement - the trait is more consistent with the rest of `rust-rocksdb`. Let me know if you disagree!

pcholakov replied:
PS. This should be:

```suggestion
    fn add_table_properties_collector_factory<F>(&mut self, factory: F)
```

Missed cleaning that up.

pcholakov commented on src/table_properties.rs:
```diff
@@ -0,0 +1,244 @@
+    fn get_readable_properties(&self) -> HashMap<CString, CString>;
+
+    /// Name of the collector to use for logging
+    fn name(&self) -> &CStr;
```
Another potential API simplification is to reuse the factory name here, and let the callback handler provide the name implementation directly. Let me know if you think this would be better; I'm more inclined to keep it dumb and aligned with the C++ interface.

AhmedSoliman replied:
I'd suggest we return &'str instead and unwrap the error of it's not a utf-8 string. I don't think the high-level rust wrapper should expose `CStr` imho.

It'd probably be best if this returns `'static &str` to make it easy to pass it down without worrying about the backing allocation.

pcholakov replied:
I like the suggestion - I had something similar at one point - just wanted to call out that there's precedent in `CompactionFilterFactory`/`CompactionFilter` to return `CStr` from these lower-level callback traits :-)

I'll play around with `'static &str` some more but then we'd still need to repeat the C-string conversion, and that gets thrown away after the call.

pcholakov replied:
I played around with this a bit more, and I don't think a return type of `&'static str` is a good idea. The problem is in `TablePropertiesCollectorCallback::name`, which is called by back by the FFI layer. It's signature makes sense to be something like:

```rs
extern "C" fn name(...) -> *const c_char;
```

so that we can conveniently bridge it to the C++ `TablePropertiesCollector::Name()` definition which is:

```cpp
virtual const char* Name() const = 0;
```

Returning a Rust `str` here requires conversion to a C-style string first, which would have to be owned by `TablePropertiesCollectorCallback`. Even if the returned `str` has a `'static` lifetime, that doesn't help us because it's a `CStr` that we need to return to RocksDB. Conversely, returning a `CStr` directly makes everything the entire call chain much simpler. I also think that making this `'static` would be over-constraining - it's ok to return a `CStr` owned by the struct itself, it doesn't need to outlive the collector. And it's still pretty convenient to have a static literal, the simplest implementation is just:

```rs
fn name(&_self) -> &CStr { c"collector_name" }
```

So bottom line: I'm fairly sure the current signature is pretty good! But please tell me if you think I'm missing something.

AhmedSoliman replied:
I don't mind any approach you take, but building an idiomatic rust interface should not include exposing `&CStr` especially if this particular string is guaranteed to be well formed utf-8 because it started its life as a rust string.

pcholakov replied:
I will concede it's uncommon but I'd suggest that for implementing a Rust callback that's going to be exposed to C, the current signature is as idiomatic as it gets. To me the `CStr` type better communicates the requirement that this *not* any normal Unicode string, but a nul-terminated C string. Open to doing it another way, of course, but I genuinely believe this fits better with the existing patterns in `rust-rocksdb` as well as the intended use.

AhmedSoliman replied:
Ok, I think I agree with you on this one.

pcholakov reviewed (commented):
Thank you so much for the UTF8 catch!

I think this is in reasonably good shape for another pass now.

AhmedSoliman commented on src/table_properties.rs:
```diff
@@ -0,0 +1,248 @@
+        )
+    }
+
+    unsafe extern "C" fn name(raw_self: *mut c_void) -> *const c_char {
+        let factory = &*(raw_self.cast_const() as *const F);
+        factory.name().as_ptr()
+    }
```
I didn't trace the entire flow, but can you verify that that we are not leaking strings somewhere, I remember seeing `strdup` somewhere in the C binding re. to names (not sure which) and I just want to make sure that this allocated string is eventually owned and deallocated.

pcholakov replied:
I think you're thinking of this:  [`rocksdb_table_properties_get_user_collected_property`](https://github.com/restatedev/rocksdb/blob/ab8bd8455a055c26885c5f16734c816d379dc669/db/c.cc#L499-L507) which is used for getting values out of the map (see the comment against `get_user_collected_property` below).

I put a lot of faith in running the tests under Valgrind - it's caught all sorts of issues along the way. If anything still remains, it's only because I've missed some paths in my tests, but I am fairly confident how things stand.

pcholakov commented:
Ready for review! I'd like to do some light benchmarking - there are a few concerns I'd like to cover like make sure the `add_user_key` doesn't bottleneck flushing unnecessarily, and building up the string property maps.

@tillrohrmann sending this one to you, cc: @AhmedSoliman for visibility. I've linked the `rust-rocksdb` PRs this depends on in the description.

pcholakov commented on src/event_listener.rs:
```diff
@@ -0,0 +1,238 @@
+            let value_string = CStr::from_ptr(value_ptr).to_owned();
+            ffi::rocksdb_free(value_ptr as *mut c_void);
+            Some(value_string)
+        }
```
Here, `value_ptr` points to a `strdup`'d string that's later freed using `rocksdb_free`. It wasn't initially apparent to me that I couldn't just return the `c_str()` pointer to the underlying string. The reason why we need to `strdup` it - and therefore free it here - is that it doesn't remain valid once the map iterator gets destroyed.

pcholakov committed:
get_user_collected_property accepts CStrLike instead of &str

pcholakov commented on src/event_listener.rs:
```diff
@@ -137,6 +146,34 @@ pub enum FlushReason {
+}
+
+impl<'a> FlushJobInfo<'a> {
+    pub fn get_user_collected_property(&self, key: impl CStrLike) -> Option<CString> {
```
`CStrLike` is a trait provided by `rust-rocksdb` - it allows the user to use any of `&str` | `&String` | `&CStr` | `&CString` when calling the API. Unfortunately lifetime requirements mean that this doesn't work great as a return type for `name()` elsewhere.

pcholakov reviewed (commented):
Thanks for the input, @AhmedSoliman!

I played around with the return type of the `name()` method but found that I still preferred the current version - the existing `CompactionFilter`/`CompactionFilterFactory` interfaces also use this type, I guess for much the same reason: it's a good match to the C++ interface and keeps the FFI bridge simple.

I did make one small improvement to `get_user_collected_property` to support more parameters as inputs, but it still also returns a `CString`.

tillrohrmann commented on src/event_listener.rs:
```diff
@@ -137,6 +146,34 @@ pub enum FlushReason {
+    pub fn get_user_collected_property(&self, key: impl CStrLike) -> Option<CString> {
+        unsafe {
+            let key_cstring = key.into_c_string().unwrap();
+            let value_ptr = ffi::rocksdb_table_properties_get_user_collected_property(
+                self.table_properties.as_ptr(),
+                key_cstring.as_ptr(),
+            );
+
+            if value_ptr.is_null() {
+                return None;
+            };
+
+            let value_string = CStr::from_ptr(value_ptr).to_owned();
+            ffi::rocksdb_free(value_ptr as *mut c_void);
+            Some(value_string)
```
What's the practice to decide whether a ffi function returns an owned or a borrowed string? There are a few functions like `rocksdb_flushjobinfo_cf_name` which seem to return a borrowed string. Looking at the usage, it seems that a borrowed string value could have sufficed as well since we are converting it into an owned Rust string and then immediately free the owned value via `rocksdb_free`.

pcholakov replied:
Generally, always favor returning a reference, as long as it lives long enough. In this case, I could not find a way to do so without a `strdup` on the C side. This is the implementation we're calling:

https://github.com/restatedev/rocksdb/pull/8/commits/805888c4babad622aee85dceb7518ffaeca4b010#diff-a98eedc10ccc709205c834eb93b247d508a5ece419799b6ea69ba16107c1a0f4R499-R507

Although we manage the lifetime of `rocksdb_table_properties_t`, and the underlying C++ `std::map` lives long enough, it's not safe to return a pointer to the string's bytes across the FFI boundary. Even deallocating the iterator that was used to find the matching map entry seems to invalidate the reference, so we have to `strdup` it to be safe. And once we have done that, we have to later free it. A possible improvement to this interface is to return a `&CStr` would the Rust caller to avoid a third copy depending on the use case.

There are many precedents for how this works, here's just one:

C FFI `rocksdb_property_value`: https://github.com/facebook/rocksdb/blob/ab8bd8455a055c26885c5f16734c816d379dc669/db/c.cc#L2069-L2077

And one of the Rust callsites: https://github.com/facebook/rocksdb/blob/ab8bd8455a055c26885c5f16734c816d379dc669/db/c.cc#L2069-L2077

tillrohrmann commented on src/table_properties.rs:
```diff
@@ -0,0 +1,244 @@
+    /// Returns human-readable properties used for logging
+    ///
+    /// It will only be called after finish() has been called.
+    fn get_readable_properties(&self) -> HashMap<CString, CString>;
```
Could this be a borrowed return value? Then the implementation might not have to allocate a `HashMap` on every call.

AhmedSoliman replied:
I believe that all strings that we expose in the public APIs should be rust's `String`, or a Vec<u8>/Box<[u8]> if the string is just bag of bytes (common in C++). I'm not convinced that returning CString to user code makes a lot of sense.

pcholakov replied:
Done!

tillrohrmann commented on src/table_properties.rs:
```diff
@@ -0,0 +1,244 @@
+        user_collected_properties: *mut ffi::rocksdb_user_collected_properties_t,
+    ) {
+        let collector: &mut C = &mut *(raw_collector.cast());
+        let props = collector.get_readable_properties();
```
Could this return an iterator? Then we wouldn't have to allocate a `HashMap` here.

pcholakov replied:
Done!

tillrohrmann commented on src/table_properties.rs:
```diff
@@ -0,0 +1,244 @@
+    ) -> bool {
+        let collector: &mut C = &mut *(raw_collector.cast());
+
+        let mut props = HashMap::new();
+        collector.finish(&mut props);
```
Same here. If `finish` returned an iterator, then we wouldn't have to allocate a `HashMap`.

tillrohrmann replied:
Theoretically, the Rust `TablePropertiesCollector` wouldn't have to return a `CString` but the conversion could happen here in the callback.

pcholakov replied:
I like the suggestion; I couldn't make it work with both owned or referenced `CString`s but both `finish` and `get_readable_properties` properties can now return an iterator over something that resolves to a `&CStr`. This allows us to stash and reuse the same properties for both if desired.

tillrohrmann reviewed (commented):
Thanks for creating this PR @pcholakov. I left a few minor comments/questions.

pcholakov committed:
Update TablePropertiesCollector to use iterators

Replaces the requirement to create a HashMap with a friendlier iterator
over generic CStr-friendly references.

pcholakov commented on src/table_properties.rs:
```diff
@@ -0,0 +1,272 @@
+    fn add_table_properties_collector_factory<F, K, V>(&mut self, factory: F)
+    where
+        F: TablePropertiesCollectorFactory<K, V> + Send + 'static,
+        K: AsRef<CStr>,
+        V: AsRef<CStr>;
```
See the `TablePropertiesCollector` trait for more.

pcholakov commented on src/table_properties.rs:
```diff
@@ -0,0 +1,272 @@
+}
+
+/// Table properties collector trait
+pub trait TablePropertiesCollector<K, V>
+where
+    K: AsRef<CStr>,
+    V: AsRef<CStr>,
```
I was hoping to make this something more like `Borrow<CStrLike>` but hit multiple issues making the generics and ownership/references work. As it stands, this allows the implementation to return an `Iter` or a collection like `Vec` which implements `IntoIter` over a bag of `CStr`s or `CString` references. I think I can easily make this take `&str` as well but then the error handling of Rust to C string conversion is buried in the callback shim, and I really don't like that.

pcholakov reviewed (commented):
Thanks again @tillrohrmann - I've added a basic iterator approach that doesn't require constructing two maps on every property collection. It probably still requires allocating something but the updated design at least allows the stored and user-readable properties to be shared.

I'll take another self-review pass so this might have some niggles still, but is ready for another review if someone has the cycles.

pcholakov committed:
Simplify property collection return types

pcholakov committed:
Remove use of rocksdb::Error in properties trait

This type cannot be constructed by library users so is not helpful here. We may model a custom error type on RocksDB's Status in the future.

pcholakov commented on src/table_properties.rs:
```diff
@@ -0,0 +1,246 @@
+        entry_type: EntryType,
+        seq: u64,
+        file_size: u64,
+    ) -> Result<(), ()>;
```
I was hoping to use the crate's own `rocksdb::Error` here but construction is meant to be private to the crate. It seems way out of scope to model RocksDB's `Status` as a Rust error type - particularly given how RocksDB's usage of these methods amounts to OK/not-OK. What's a good course of action here? I want `add_user_key` to be consistent with `finish` and return `Result`, but it also feels awkward returning `Result<_, ()>`.

pcholakov committed:
Introduce CollectorError, update trait signatures

pcholakov committed:
Add FlushJobInfo::get_user_collected_property_keys

pcholakov commented:
@AhmedSoliman updated to take advantage of https://github.com/restatedev/rocksdb/pull/9

pcholakov committed:
Add user property key filtering support

AhmedSoliman commented on src/event_listener.rs:
```diff
@@ -137,6 +146,60 @@ pub enum FlushReason {
+        }
+    }
+
+    pub fn get_user_collected_property_keys(&self, prefix: Option<impl CStrLike>) -> Vec<&CStr> {
```
potentially overkill, an impl CStrLike should be enough and users who don't want prefix can pass `""` akin to what you are doing internally really.

AhmedSoliman commented on src/event_listener.rs:
```diff
@@ -137,6 +146,60 @@ pub enum FlushReason {
+    pub fn get_user_collected_property_keys(&self, prefix: Option<impl CStrLike>) -> Vec<&CStr> {
+        unsafe {
+            let mut key_count: usize = 0;
+            let prefix = prefix.map_or_else(|| c"".to_owned(), |p| p.into_c_string().unwrap());
```
just FYI, `String::new()` is a `const fn` which means you don't need a closure nor allocating every time.

AhmedSoliman commented on src/event_listener.rs:
```diff
@@ -137,6 +146,60 @@ pub enum FlushReason {
+                    result.push(CStr::from_ptr(key_ptr));
+                }
+            }
+            ffi::rocksdb_free(keys_ptr as *mut c_void);
```
👏

AhmedSoliman commented on src/table_properties.rs:
```diff
@@ -0,0 +1,264 @@
+
+        let Ok(props) = collector.finish() else {
+            // An error will be logged by RocksDB to its own log, though the details will be swallowed.
+            // Property collectors should perform their own logging via other mehanisms if required.
```
```suggestion
            // Property collectors should perform their own logging via other mechanisms if required.
```

pcholakov replied:
Oops, reinstalled spell checker :-) Thanks!

AhmedSoliman reviewed (approved):
Looks good. Left a couple of nits for your consideration.

pcholakov committed:
Addressing PR comments
